File: src/game.py
import pygame
import sys
from src.level import Level
from src.screen import draw_game
from src.image import chop_into_frames
from src.player import Player
from src.fish import Fish
from src.sudden_death import SuddenDeath

pygame.init()

SCALE_FACTOR = 1  # Scale factor to enlarge the window
window = None
pygame.display.set_caption("TPC3")
base_resolution = None
center = None
# Create a counter for the points of each team
team_points = [0, 0]
# Create a list to check if a team has died
death = [False, False]
# Get team names
team_names = ["Pablo's cool long team name", "Dani"]
# Create the text font that counts each teams points
font = pygame.font.Font(None, 36)
# Create a fixed-resolution surface
base_surface = None

# Load images
tileset = None
background = None
fish_sprites = None
sudden_death_sprites = None
# Create Level
current_level = None

# Create Sudden Death
sudden_death = None

# Create sharks
player_a = None
player_b = None

# Create fish
fish = None

# Create Controllers
controller_1 = None
controller_2 = None

a_holding = False
b_holding = False


def set_up(map_file, player1, player2, full_screen=False):
    global base_resolution, center, window, base_surface, player_a, player_b, fish, current_level, controller_1, controller_2, a_holding, b_holding, death, sudden_death
    global tileset, background, fish_sprites, sudden_death_sprites

    # Read map
    current_level = Level(map_file)

    # Set up screen
    base_resolution = (current_level.get_width() * 32, current_level.get_height() * 32)
    center = (base_resolution[0] / 2, base_resolution[1] / 2)
    # window = pygame.display.set_mode((base_resolution[0] * SCALE_FACTOR, base_resolution[1] * SCALE_FACTOR), pygame.RESIZABLE) 
    
    if not full_screen:
        window = pygame.display.set_mode((800, 600), pygame.RESIZABLE)
    elif window is None:
        window = pygame.display.set_mode((0, 0), pygame.FULLSCREEN)
    
    if not tileset:
        tileset = chop_into_frames(pygame.image.load("./assets/tileset.png").convert_alpha(), 32 , 32)
        background = chop_into_frames(pygame.image.load("./assets/background.png").convert_alpha(), 32 , 32)
        fish_sprites = chop_into_frames(pygame.image.load("./assets/Pufferfish.png").convert_alpha(), 32, 32)
        sudden_death_sprites = chop_into_frames(pygame.image.load("./assets/SuddenDeath.png").convert_alpha(), 32, 32)
    
    
    base_surface = pygame.Surface(base_resolution)
    window.fill((0, 0, 0))
    pygame.display.flip()

    controller_1 = player1
    controller_2 = player2
    controller_1.info()
    controller_2.info()
    controller_1.set_is_first_controller(True)
    controller_2.set_is_first_controller(False)
    controller_1.set_level(current_level.get_matrix())
    controller_2.set_level(current_level.get_matrix())
    controller_1.make_first_time_pixel_map()
    controller_2.make_first_time_pixel_map()
    team_names[0] = controller_1.get_team_name()
    team_names[1] = controller_2.get_team_name()

    a_pos = current_level.get_a_starting_pos()
    b_pos = current_level.get_b_starting_pos()
    player_a = Player(a_pos[0], a_pos[1], controller_1.get_look(), controller_1.get_color(), current_level.get_matrix())
    player_b = Player(b_pos[0], b_pos[1], controller_2.get_look(), controller_2.get_color(), current_level.get_matrix())

    fish = []
    for pos in current_level.get_fishes_starting_pos():
        fish.append(Fish(pos, current_level.get_matrix()))

    sudden_death = SuddenDeath(current_level.get_width(), current_level.get_height())


    a_holding = False
    b_holding = False

    death[0] = False
    death[1] = False


def loop():
    global current_level, a_holding, b_holding, death, sudden_death

    photo_finish_time = 0
    game_time = 0

    # Main loop
    while True:
        # Handle events
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()

        # <~~~~Logic Block~~~~>
        if (not death[0] and not death[1]):
            fish_pos = []
            fish_state = []
            for current in fish:
                fish_pos.append(current.get_pos())
                fish_state.append(current.get_state())
            
            controller_1.clear()
            controller_2.clear()
            controller_1.update(fish_pos, fish_state, player_a.get_pos(), player_b.get_pos(), game_time // 4, player_a.get_y_speed())
            controller_2.update(fish_pos, fish_state, player_b.get_pos(), player_a.get_pos(), game_time // 4, player_b.get_y_speed())
            controller_1.behavior()
            controller_2.behavior()

            player_a.control(controller_1.get_go_left(), controller_1.get_go_right(), controller_1.get_jump())
            player_b.control(controller_2.get_go_left(), controller_2.get_go_right(), controller_2.get_jump())

            # Throws
            if (controller_1.get_throw_down() and a_holding):
                player_a.fish_jump()
                a_holding = False
            if (controller_2.get_throw_down() and b_holding):
                player_b.fish_jump()
                b_holding = False
            if ((controller_1.get_throw_right() or controller_1.get_throw_left()) and a_holding):
                a_holding = False
            if ((controller_2.get_throw_right() or controller_2.get_throw_left()) and b_holding):
                b_holding = False

            for current in fish:
                if (current.behave(player_a.get_pos(), player_b.get_pos(), player_a.get_facing(), player_b.get_facing(),
                                    controller_1.get_grab() and not a_holding, controller_2.get_grab() and not b_holding,
                                    controller_1.get_throw_down(), controller_1.get_throw_right(), controller_1.get_throw_left(),
                                    controller_2.get_throw_down(), controller_2.get_throw_right(), controller_2.get_throw_left())):
                    death[0] = True
                    death[1] = True
                    team_points[0] += 1
                    team_points[1] += 1

                if (current.is_grabbed_by_a()):
                    a_holding = True
                if (current.is_grabbed_by_b()):
                    b_holding = True
                
                if (current.is_hit(player_a.get_pos())):
                    death[0] = True
                    team_points[1] += 1
                if (current.is_hit(player_b.get_pos())):
                    death[1] = True
                    team_points[0] += 1
            if (sudden_death.check_death(game_time // 4, player_a.get_pos())):
                death[0] = True
                team_points[1] += 1
            if (sudden_death.check_death(game_time // 4, player_b.get_pos())):
                death[1] = True
                team_points[0] += 1
            game_time += 1
        else:
            photo_finish_time += 1
            if (photo_finish_time == 60):
                break

        # <~~~~Draw Block~~~~>
        # Draw background
        current_level.draw_background(background, base_surface, 1, 0, 0)
        # Draw map
        current_level.draw(tileset, base_surface, 1, 0, 0)
        # Draw Fishes
        # This part of the code is horrible I know. I just have to finish this project on time and refactoring would take a while
        for current in fish:
            current.draw(base_surface, fish_sprites, player_a.get_facing(), player_b.get_facing())

        # Draw players
        player_a.draw(base_surface)
        player_b.draw(base_surface)

        # Draw Sudden Death
        sudden_death.draw_spikes(base_surface, sudden_death_sprites, game_time // 4)

        # Draw logos
        # base_surface.blit(logos, (base_resolution[0]-270, 20))

        # Draw text of points
        text_of_points = font.render(team_names[0] + ": " + str(team_points[0]) + " - " + team_names[1] + ": " + str(team_points[1]), True, (255, 255, 255))
        base_surface.blit(text_of_points, (4, 2))

        draw_game(window, base_surface, current_level)

        # Limit to 60 FPS
        pygame.time.Clock().tick(60)

File: src/Game.cpp
#include "Game.hpp"
#include <iostream>
#include <filesystem>
#include <array>

// Font paths to try (common on Linux)
static const char* FONT_PATHS[] = {
    "/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf",
    "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
    "/usr/share/fonts/truetype/freefont/FreeSans.ttf",
    "/usr/share/fonts/truetype/ubuntu/Ubuntu-R.ttf",
    // Common Windows font fallback
    "C:\\Windows\\Fonts\\arial.ttf",
    // Local assets fallback (if you bundle a font into assets/)
    "./assets/DejaVuSans.ttf",
    nullptr
};

Game::Game() {
    for (int i = 0; FONT_PATHS[i]; ++i) {
        if (std::filesystem::exists(FONT_PATHS[i])) {
            if (hud_font.loadFromFile(FONT_PATHS[i])) {
                font_loaded = true;
                break;
            }
        }
    }
}

void Game::load_textures() {
    if (textures_loaded) return;
    tileset_frames = load_frames("./assets/tileset.png",    tileset_tex, 32, 32);
    bg_frames      = load_frames("./assets/background.png", bg_tex,      32, 32);
    fish_frames    = load_frames("./assets/Pufferfish.png", fish_tex,    32, 32);
    sd_frames      = load_frames("./assets/SuddenDeath.png",sd_tex,      32, 32);
    textures_loaded = true;
}

void Game::set_up(const std::string& map_file,
                  std::unique_ptr<Controller> p1,
                  std::unique_ptr<Controller> p2,
                  bool full_screen) {
    load_textures();

    current_level = std::make_unique<Level>(map_file);

    int lw = current_level->get_width()  * 32;
    int lh = current_level->get_height() * 32;
    base_surface.create(lw, lh);

    // Create or reuse window
    if (!window.isOpen()) {
        if (full_screen)
            window.create(sf::VideoMode::getDesktopMode(), "TPC3",
                          sf::Style::Fullscreen);
        else
            window.create(sf::VideoMode(800, 600), "TPC3", sf::Style::Resize);
        window.setFramerateLimit(60);
    }

    controller_1 = std::move(p1);
    controller_2 = std::move(p2);

    controller_1->set_is_first_controller(true);
    controller_2->set_is_first_controller(false);
    controller_1->set_level(current_level->get_matrix());
    controller_2->set_level(current_level->get_matrix());
    controller_1->make_first_time_pixel_map();
    controller_2->make_first_time_pixel_map();
    team_names[0] = controller_1->get_team_name();
    team_names[1] = controller_2->get_team_name();

    auto a_pos = current_level->get_a_starting_pos();
    auto b_pos = current_level->get_b_starting_pos();
    player_a = std::make_unique<Player>(a_pos.first, a_pos.second,
                                        controller_1->get_look(),
                                        controller_1->get_color(),
                                        current_level->get_matrix());
    player_b = std::make_unique<Player>(b_pos.first, b_pos.second,
                                        controller_2->get_look(),
                                        controller_2->get_color(),
                                        current_level->get_matrix());

    fish.clear();
    for (auto& fp : current_level->get_fishes_starting_pos())
        fish.emplace_back(fp, current_level->get_matrix());

    sudden_death = std::make_unique<SuddenDeath>(
        current_level->get_width(), current_level->get_height());

    a_holding  = false;
    b_holding  = false;
    death      = {false, false};
    game_time  = 0;
    frame_rate = 60;
}

void Game::loop() {
    sf::Clock clock;
    int photo_finish = 0;

    while (true) {
        // --- Events ---
        sf::Event ev;
        while (window.pollEvent(ev)) {
            if (ev.type == sf::Event::Closed) {
                window.close();
                std::cout << "Final scores: [" << points[0] << ", " << points[1] << "]\n";
                return;
            }
        }

        // --- Speed control ---
        static constexpr int FPS_NORMAL = 60;
        static constexpr int FPS_X2     = 120;
        static constexpr int FPS_X4     = 240;
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::Num1)) frame_rate = FPS_NORMAL;
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::Num2)) frame_rate = FPS_X2;
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::Num3)) frame_rate = FPS_X4;

        // --- Logic ---
        if (!death[0] && !death[1]) {
            std::vector<std::pair<int,int>> fp;
            std::vector<int>               fs;
            for (auto& f : fish) { fp.push_back(f.get_pos()); fs.push_back(f.get_state()); }

            controller_1->clear();
            controller_2->clear();
            controller_1->update(fp, fs, player_a->get_pos(), player_b->get_pos(), game_time, player_a->get_y_speed());
            controller_2->update(fp, fs, player_b->get_pos(), player_a->get_pos(), game_time, player_b->get_y_speed());
            controller_1->behavior();
            controller_2->behavior();

            player_a->control(controller_1->get_go_left(),
                              controller_1->get_go_right(),
                              controller_1->get_jump());
            player_b->control(controller_2->get_go_left(),
                              controller_2->get_go_right(),
                              controller_2->get_jump());

            // Throws
            if (controller_1->get_throw_down()  && a_holding) { player_a->fish_jump(); a_holding = false; }
            if (controller_2->get_throw_down()  && b_holding) { player_b->fish_jump(); b_holding = false; }
            if ((controller_1->get_throw_right() || controller_1->get_throw_left()) && a_holding) a_holding = false;
            if ((controller_2->get_throw_right() || controller_2->get_throw_left()) && b_holding) b_holding = false;

            for (auto& f : fish) {
                bool tie = f.behave(
                    player_a->get_pos(), player_b->get_pos(),
                    player_a->get_facing(), player_b->get_facing(),
                    controller_1->get_grab() && !a_holding,
                    controller_2->get_grab() && !b_holding,
                    controller_1->get_throw_down(),
                    controller_1->get_throw_right(),
                    controller_1->get_throw_left(),
                    controller_2->get_throw_down(),
                    controller_2->get_throw_right(),
                    controller_2->get_throw_left());

                if (tie) { death[0] = death[1] = true; ++points[0]; ++points[1]; }
                if (f.is_grabbed_by_a()) a_holding = true;
                if (f.is_grabbed_by_b()) b_holding = true;
                if (f.is_hit(player_a->get_pos())) { death[0] = true; ++points[1]; }
                if (f.is_hit(player_b->get_pos())) { death[1] = true; ++points[0]; }
            }
            if (sudden_death->check_death(game_time, player_a->get_pos())) { death[0] = true; ++points[1]; }
            if (sudden_death->check_death(game_time, player_b->get_pos())) { death[1] = true; ++points[0]; }

            ++game_time;
        } else {
            ++photo_finish;
            if (photo_finish == 60) break;
        }

        // --- Draw ---
        draw_frame();
        scale_and_present();

        // --- FPS limit ---
        sf::Time elapsed = clock.restart();
        sf::Time target  = sf::seconds(1.f / frame_rate);
        if (elapsed < target) sf::sleep(target - elapsed);
    }
}

void Game::draw_frame() {
    base_surface.clear(sf::Color::Black);
    current_level->draw_background(bg_tex,      bg_frames,      base_surface, 1.f, 0.f, 0.f);
    current_level->draw           (tileset_tex,  tileset_frames, base_surface, 1.f, 0.f, 0.f);
    for (auto& f : fish)
        f.draw(base_surface, fish_tex, fish_frames,
               player_a->get_facing(), player_b->get_facing());
    player_a->draw(base_surface);
    player_b->draw(base_surface);
    sudden_death->draw_spikes(base_surface, sd_tex, sd_frames, game_time);

    // HUD text
    if (font_loaded) {
        std::string hud = team_names[0] + ": " + std::to_string(points[0])
                        + " - " + team_names[1] + ": " + std::to_string(points[1]);
        sf::Text txt(hud, hud_font, 14u);
        txt.setFillColor(sf::Color::White);
        txt.setPosition(4.f, 2.f);
        base_surface.draw(txt);
    }

    base_surface.display();
}

void Game::scale_and_present() {
    sf::Vector2u ws  = window.getSize();
    sf::Vector2u bs  = base_surface.getSize();
    float w_ratio    = (float)ws.x / ws.y;
    float b_ratio    = (float)bs.x / bs.y;
    float scale;
    float xo, yo;
    if (w_ratio > b_ratio) {
        scale = (float)ws.y / bs.y;
        xo    = (ws.x - bs.x * scale) / 2.f;
        yo    = 0.f;
    } else {
        scale = (float)ws.x / bs.x;
        xo    = 0.f;
        yo    = (ws.y - bs.y * scale) / 2.f;
    }
    sf::Sprite sp(base_surface.getTexture());
    sp.setScale(scale, scale);
    sp.setPosition(xo, yo);
    window.clear(sf::Color::Black);
    window.draw(sp);
    window.display();
}
